android/app/src/main/java/org/saltyrtc/demo/app/webrtc/PeerConnectionHelper.java

```diff
@@ -0,0 +1,389 @@
+                ufcdc.write(new DataChannel.Buffer(message, true));
+            }
+        };
+
+        // Bind events
+        dc.registerObserver(new DataChannel.Observer() {
+            @Override
+            public void onBufferedAmountChange(final long bufferedAmount) {
+            	ufcdc.bufferedAmountChange();
+			}
+
+            @Override
+            public void onStateChange() {
+                switch (dc.state()) {
+                    case CONNECTING:
+                        log.debug("Data channel " + dc.label() + " connecting");
+                        break;
+                    case OPEN:
+                        if (PeerConnectionHelper.this.dcOpened) {
+                            log.error("Data channel " + dc.label() + " re-opened");
+                        } else {
+                            PeerConnectionHelper.this.dcOpened = true;
+                            log.info("Data channel " + dc.label() + " open");
+                            task.handover(handler);
+                        }
+                        break;
+                    case CLOSING:
+                        if (!PeerConnectionHelper.this.dcOpened) {
+                            log.error("Data channel " + dc.label() + " closing");
+                        } else {
+                            log.debug("Data channel " + dc.label() + " closing");
+                            try {
+                                link.closing();
+                            } catch (UntiedException error) {
+                                log.warn("Could not move into closing state", error);
+                            }
+                        }
+                        break;
+                    case CLOSED:
+                        if (!PeerConnectionHelper.this.dcOpened) {
+                            log.error("Data channel " + dc.label() + " closed");
+                        } else {
+                            log.info("Data channel " + dc.label() + " closed");
+                            try {
+                                link.closed();
+                            } catch (UntiedException error) {
+                                // Note: We can safely ignore this because, in
+								//       our case, the signalling instance may
+								//       be closed before the channel has been
+								//       through the closing sequence.
+                            }
+                        }
+						dc.dispose();
+                        break;
+                }
+            }
+
+            @Override
+            public void onMessage(@NonNull final DataChannel.Buffer buffer) {
+                if (!buffer.binary) {
+                    task.close(CloseCode.PROTOCOL_ERROR);
+                } else {
+                    try {
+                        link.receive(buffer.data);
+                    } catch (UntiedException error) {
+                        log.warn("Could not feed incoming data", error);
```

Comment:
That message is a bit unclear. Is the data hungry? :slightly_smiling_face: 

Maybe this is better?

```suggestion
                        log.warn("Could not feed incoming data to the transport link", error);
```